README.md
# JSON API for Dataset Arte Electrónico Argentino

A very simple RESTful JSON API to query the dataset curated by [@cristianReynaga](https://github.com/cristianReynaga), you can read more about it [here](https://github.com/cristianReynaga/Dataset-Arte-Electronico-Argentino) (in spanish).  
This API is built using Golang and SQLite3. I also provide a sanitised csv file ready to be imported to a sqlite db (with the necessary statements to ease the process).  
  
This is a WIP with the intention to have a quick way to access this information and to keep learning golang :-)  
It's publicly accessible here (For now `http://` only): 

* [website](http://arte-electronico.cyberpunk.com.ar/)  
* [API](http://api.arte-electronico.cyberpunk.com.ar/)  

## How to ...?

### Easy way

```bash
$ docker pull lvm23/api-arte-electronico-argentino
$ docker run -d -p 8080:8080 lvm23/api-arte-electronico-argentino
```

### Manual way

As usual

```bash
$ git clone https://github.com/lvm/api-dataset-arte-electronico-argentino
$ cd api-dataset-arte-electronico-argentino/
```

Then, create the SQLite database  

```bash
$ rm -f db/electronicArtArgentina.sqlite3
$ cd data
$ cat csv-to-sqlite | sqlite3 ../db/electronicArtArgentina.sqlite3
```

For testing purpose, to build the API (I recommend using the [official golang image](https://hub.docker.com/_/golang/)) you need to execute

```bash
$ export GOPATH=$HOME
$ go get .
$ go run
```

Note: `api.go` reads a couple of extra enviroment vars: `HTTP_PORT` (default: `8080`), `DB_FILENAME` (default: `./db/electronicArtArgentina.sqlite3`))

If everything went OK, you should see something like

```
[GIN-debug] [WARNING] Running in "debug" mode. Switch to "release" mode in production.
 - using env:   export GIN_MODE=release
 - using code:  gin.SetMode(gin.ReleaseMode)

[GIN-debug] GET    //exhibitions       --> main.GetExhibitions (4 handlers)
[GIN-debug] GET    /exhibitions/:e_id --> main.GetExhibition (4 handlers)
[GIN-debug] GET    /search            --> main.SearchExhibitions (4 handlers)
[GIN-debug] OPTIONS /exhibitions       --> main.EndpointsOptions (4 handlers)
[GIN-debug] OPTIONS /exhibitions/:e_id --> main.EndpointsOptions (4 handlers)
[GIN-debug] OPTIONS /search            --> main.EndpointsOptions (4 handlers)
[GIN-debug] Listening and serving HTTP on :8080
```

and you can start querying the API.  
Note: If you export the `GIN_MODE=release` env var, those *debug* messages won't be displayed but it's recommended for production.  
Also, you can take a look at the [Dockerfile](Dockerfile) to get an idea of the steps required.

### Endpoints

For the moment, there are two endpoints: `exhibitions` and `search`. It accepts only `GET` and `OPTION` requests.

#### `endpoints`

**List all available endpoints**
```
GET /
```

#### `exhibitions`

This method allows to query all of them or just one.  

**List all (around 15xx) exhibitions**
```
GET /exhibitions
```

To retrieve more items, use the `GET` parameter `?page=n`

**Just one by passing the ID**
```
GET /exhibitions/23
```

#### `search`

This method allows to search, primarily by the exhibition's name and then you can pass more parameters to narrow your search. These extra parameters use the `AND` operator when building the query.  

| param       | values                                                                                               |
| ----------- | ---------------------------------------------------------------------------------------------------- |
| `q`         | The full name or a part of the name of an exhibition. Can be any `string`                            |
| `year`      | The *valid* format would be `YYYY`, but you can pass any `int (32)`                                  |
| `when`      | Only takes `since` and `until`. They work as `greater-than-equal` and `less-than-equal` respectively |
| `technique` | The name of a technique, can be any `string`.                                                        |
| `curator`   | The name of a curator, can be any `string`.                                                          |
| `artist`    | The name of an artist, can be any `string`.                                                          |
| `work`      | The name of the work, can be any `string`.                                                           |
| `page`      | The page #. Each page shows 20 items.                                                                |

**Examples**

**List all exhibitions with `festival` on its name**
```
GET /search?q=festival
```

**...during 1998**
```
GET /search?q=festival&year=1998
```

**...until 1998**
```
GET /search?q=festival&year=1998&when=until
```

**...since 1998**
```
GET /search?q=festival&year=1998&when=since
```

**...with technique `videoart`**
```
GET /search?q=festival&year=1998&when=since&technique=videoart
```

**...curated by Graciela Taquini**
```
GET /search?q=festival&year=1998&when=since&technique=videoart&curator=taquini
```

**...artist Joanna Rytel**
```
GET /search?q=festival&year=1998&when=since&technique=videoart&curator=taquini&artist=rytel
```

**...with the word `sheep` on its name**
```
GET /search?q=festival&year=1998&when=since&technique=videoart&curator=taquini&artist=rytel&work=sheep
```


## TODO

* Pagination: cleaner way to create the `paginationLinks`
* Pagination: Optimize SQLite (drop `OFFSET` and `LIMIT`, use `INDEX`)
* Search by person (artists or curators)
* configure `https://`

## Bugs, contributions

Go [here](https://github.com/lvm/api-dataset-arte-electronico-argentino/issues)

## LICENSE

See [LICENSE](LICENSE)


CHANGELOG
2016-07-09 - v2
	v2 Public release
	Added entrypoints: techniques, locations, events
	Improved the search results 

2016-07-06 - v1
	Updates and pagination fixes

2016-06-29 - v1
	v1 Public release


events.go
package main

import (
	_ "github.com/mattn/go-sqlite3"
	"github.com/gin-gonic/gin"
)

type (
	Event struct {
		Id    int64 `db:"id" json:"-"`
		Event string `db:"nombre_exhibicion" json:"event,omitempty"`
		Total string `db:"count(nombre_exhibicion)" json:"total,omitempty"`
	}

	ResourceEvent struct {
		Id         int64 `json:"id"`
		Type       string `json:"type"`
		Attributes Event `json:"attributes"`
	}

	ResourceEvents struct {
		Data  []ResourceEvent `json:"data"`
		Links PaginationLinks `json:"links"`
		Meta  Meta `json:"meta"`
	}
)

func newResourceEvent (attr Event) ResourceEvent {
	return ResourceEvent {
		Id: attr.Id,
		Type: "events",
		Attributes: attr,
	}
}

func newResourceEvents (events []Event, endpoint string, pg string) ResourceEvents {
	resources := make([]ResourceEvent, 0)
	for l := range events {
		resources = append(resources, newResourceEvent(events[l]))
	}

	links := PaginationLinks {
		First: paginationLink(endpoint, firstPage, "first"),
		Last: paginationLink(endpoint, lastPage, "last"),
		Prev: paginationLink(endpoint, pg, "prev"),
		Next: paginationLink(endpoint, pg, "next"),
	}

	return ResourceEvents{Data: resources, Links: links, Meta: Meta{TotalPages: "3"}}
}


// http http://localhost:8080/api/events
func GetEvents(c *gin.Context) {
	var events []Event
	page := c.DefaultQuery("page", "1")

	query := PaginateQuery(groupByQuery("SELECT id, %s, count(%s) FROM exhibiciones", "event"), page)
	_, err := dbmap.Select(&events, query)
	checkErr(err)

	if err == nil {
		content := newResourceEvents(events, "events", page)
		c.JSON(200, content)
	} else {
		jsonErr := &Error{"not_found", 404, "Not Found Error", "No event found."}
		c.JSON(404, jsonErr)
	}
}


search.go
package main

import (
	"fmt"
	"strconv"
	_ "github.com/mattn/go-sqlite3"
	"github.com/gin-gonic/gin"
)

func howManyPages(query string) (int, error) {
	count, err := dbmap.SelectInt(query)
	return int(count) / perPage, err
}

// http http://localhost:8080/api/v1/search?q=festival[&year=1998[&when=since[&technique=videoarte[&artist=rytel[&curator=taquini[&work=sheep]]]]]]
func SearchExhibitions(c *gin.Context) {
	var exhibitions []Exhibition
	q := c.Query("q")
	year := c.Query("year")
	when := c.Query("when")
	where := c.Query("where")
	curator := c.Query("curator")
	artist := c.Query("artist")
	technique := c.Query("technique")
	work := c.Query("work")
	page := c.DefaultQuery("page","1")

	and_year := ""
	if year != "" {
		y, yerr := strconv.ParseInt(year, 10, 32)
		checkErr(yerr)

		cmp := "="
		switch when {
		case "until", "UNTIL":
			cmp = "<="
		case "since", "SINCE":
			cmp = ">="
		}

		and_year = fmt.Sprintf("AND anio %s %d", cmp, y)
	}

	and_where := ""
	if where != "" {
		and_where = fmt.Sprintf("AND nombre_locacion LIKE '%%%s%%'", where)
	}

	and_work := ""
	if work != "" {
		and_work = fmt.Sprintf("AND nombre_obra LIKE '%%%s%%'", work)
	}

	and_technique := ""
	if technique != "" {
		and_technique = fmt.Sprintf("AND tecnica LIKE '%%%s%%'", technique)
	}

	and_curator := ""
	if curator != "" {
		and_curator = fmt.Sprintf("AND curadores  MATCH '*%s*'", curator)
	}

	and_artist := ""
	if artist != "" {
		and_artist = fmt.Sprintf("AND artistas MATCH '*%s*'", artist)
	}

	from_where := fmt.Sprintf(`
FROM exhibiciones
NATURAL JOIN artistas
NATURAL JOIN curadores
WHERE nombre_exhibicion LIKE '%%%s%%'
%s %s %s %s %s %s`, q, and_year, and_where, and_technique, and_work, and_artist, and_curator)

	// TODO FIX ME. sqlite doesn't have a quick way to query for # of rows AND the rows, so this:
	totalPages, perr := howManyPages(fmt.Sprintf("SELECT count(id) %s", from_where))
	if perr == nil {
		query := PaginateQuery(fmt.Sprintf("SELECT * %s", from_where), page)

		_, err := dbmap.Select(&exhibitions, query)
		checkErr(err)

		if err == nil {
			content := newResourceExhibitions(exhibitions, "search", page, strconv.Itoa(totalPages))
			c.JSON(200, content)
		} else {
			jsonErr := &Error{"not_found", 404, "Not Found Error", "Your query didn't match any exhibition."}
			c.JSON(404, jsonErr)
		}
	} else {
		jsonErr := &Error{"server_error", 500, "Server Error", "Something wrong has happened."}
		c.JSON(500, jsonErr)
	}
}


locations.go
package main

import (
	_ "github.com/mattn/go-sqlite3"
	"github.com/gin-gonic/gin"
)

type (
	Location struct {
		Id         int64 `db:"id" json:"-"`
		Location   string `db:"nombre_locacion" json:"location,omitempty"`
		Total      string `db:"count(nombre_locacion)" json:"total,omitempty"`
	}

	ResourceLocation struct {
		Id         int64 `json:"id"`
		Type       string `json:"type"`
		Attributes Location `json:"attributes"`
	}

	ResourceLocations struct {
		Data  []ResourceLocation `json:"data"`
		Links PaginationLinks `json:"links"`
		Meta  Meta `json:"meta"`
	}
)

func newResourceLocation (attr Location) ResourceLocation {
	return ResourceLocation {
		Id: attr.Id,
		Type: "locations",
		Attributes: attr,
	}
}

func newResourceLocations (locations []Location, endpoint string, pg string) ResourceLocations {
	resources := make([]ResourceLocation, 0)
	for l := range locations {
		resources = append(resources, newResourceLocation(locations[l]))
	}

	links := PaginationLinks {
		First: paginationLink(endpoint, firstPage, "first"),
		Last: paginationLink(endpoint, lastPage, "last"),
		Prev: paginationLink(endpoint, pg, "prev"),
		Next: paginationLink(endpoint, pg, "next"),
	}

	return ResourceLocations{Data: resources, Links: links, Meta: Meta{TotalPages: "1"}}
}


// http http://localhost:8080/api/locations
func GetLocations(c *gin.Context) {
	var locations []Location
	page := c.DefaultQuery("page", "1")

	query := PaginateQuery(groupByQuery("SELECT id, %s, count(%s) FROM exhibiciones", "location"), page)
	_, err := dbmap.Select(&locations, query)
	checkErr(err)

	if err == nil {
		content := newResourceLocations(locations, "locations", page)
		c.JSON(200, content)
	} else {
		jsonErr := &Error{"not_found", 404, "Not Found Error", "No location found."}
		c.JSON(404, jsonErr)
	}
}


techniques.go
package main

import (
	_ "github.com/mattn/go-sqlite3"
	"github.com/gin-gonic/gin"
)

type (
	Technique struct {
		Id        int64 `db:"id" json:"-"`
		Technique string `db:"tecnica" json:"technique,omitempty"`
		Total     string `db:"count(tecnica)" json:"total,omitempty"`
	}

	ResourceTechnique struct {
		Id         int64 `json:"id"`
		Type       string `json:"type"`
		Attributes Technique `json:"attributes"`
	}

	ResourceTechniques struct {
		Data  []ResourceTechnique `json:"data"`
		Links PaginationLinks `json:"links"`
		Meta  Meta `json:"meta"`
	}
)

func newResourceTechnique (attr Technique) ResourceTechnique {
	return ResourceTechnique {
		Id: attr.Id,
		Type: "techniques",
		Attributes: attr,
	}
}

func newResourceTechniques (techniques []Technique, endpoint string, pg string) ResourceTechniques {
	resources := make([]ResourceTechnique, 0)
	for l := range techniques {
		if techniques[l].Technique != "" {
			resources = append(resources, newResourceTechnique(techniques[l]))
		}
	}

	links := PaginationLinks {
		First: paginationLink(endpoint, firstPage, "first"),
		Last: paginationLink(endpoint, lastPage, "last"),
		Prev: paginationLink(endpoint, pg, "prev"),
		Next: paginationLink(endpoint, pg, "next"),
	}

	return ResourceTechniques{Data: resources, Links: links, Meta: Meta{TotalPages: "6"}}
}


// http http://localhost:8080/api/techniques
func GetTechniques(c *gin.Context) {
	var techniques []Technique
	page := c.DefaultQuery("page", "1")

	query := PaginateQuery(groupByQuery("SELECT id, %s, count(%s) FROM exhibiciones", "technique"), page)
	_, err := dbmap.Select(&techniques, query)
	checkErr(err)

	if err == nil {
		content := newResourceTechniques(techniques, "techniques", page)
		c.JSON(200, content)
	} else {
		jsonErr := &Error{"not_found", 404, "Not Found Error", "No technique found."}
		c.JSON(404, jsonErr)
	}
}


exhibitions.go
package main

import (
	"time"
	_ "github.com/mattn/go-sqlite3"
	"github.com/gin-gonic/gin"
)

type (
	Exhibition struct {
		Id         int64 `db:"id" json:"-"`
		Anio       int `db:"anio" json:"year,omitempty"`
		Nombre     string `db:"nombre_exhibicion" json:"exhibition,omitempty"`
		Locacion   string `db:"nombre_locacion" json:"location,omitempty"`
		Curador0   string `db:"curador_0" json:"curator_0,omitempty"`
		Curador1   string `db:"curador_1" json:"curator_1,omitempty"`
		Curador2   string `db:"curador_2" json:"curator_2,omitempty"`
		Curador3   string `db:"curador_3" json:"curator_3,omitempty"`
		FechaIncio time.Time `db:"fecha_inicio" json:"start,omitempty"`
		FechaFin   time.Time `db:"fecha_finalizacion" json:"end,omitempty"`
		Obra       string `db:"nombre_obra" json:"work,omitempty"`
		Artista0   string `db:"nombre_artista_0" json:"artist_0,omitempty"`
		Artista1   string `db:"nombre_artista_1" json:"artist_1,omitempty"`
		Artista2   string `db:"nombre_artista_2" json:"artist_2,omitempty"`
		Artista3   string `db:"nombre_artista_3" json:"artist_3,omitempty"`
		Artista4   string `db:"nombre_artista_4" json:"artist_3,omitempty"`
		Artista5   string `db:"nombre_artista_5" json:"artist_5,omitempty"`
		Artista6   string `db:"nombre_artista_6" json:"artist_6,omitempty"`
		Artista7   string `db:"nombre_artista_7" json:"artist_7,omitempty"`
		Artista8   string `db:"nombre_artista_8" json:"artist_8,omitempty"`
		Tecnica    string `db:"tecnica" json:"technique,omitempty"`
	}

	ResourceExhibition struct {
		Id         int64 `json:"id"`
		Type       string `json:"type"`
		Attributes Exhibition `json:"attributes"`
	}

	ResourceExhibitions struct {
		Data  []ResourceExhibition `json:"data"`
		Links PaginationLinks `json:"links"`
		Meta  Meta `json:"meta"`
	}
)

func newResourceExhibition (attr Exhibition) ResourceExhibition {
	return ResourceExhibition {
		Id: attr.Id,
		Type: "exhibitions",
		Attributes: attr,
	}
}

func newResourceExhibitions (exhibitions []Exhibition, endpoint string, pg string, totalPages string) ResourceExhibitions {
	resources := make([]ResourceExhibition, 0)
	for e := range exhibitions {
		resources = append(resources, newResourceExhibition(exhibitions[e]))
	}

	links := PaginationLinks {First: "",Last: "",Prev: "",Next: ""}
	if pg != "" {
		links = PaginationLinks {
			First: paginationLink(endpoint, firstPage, "first"),
			Last: paginationLink(endpoint, lastPage, "last"),
			Prev: paginationLink(endpoint, pg, "prev"),
			Next: paginationLink(endpoint, pg, "next"),
		}
	}

	if totalPages == "" {
		totalPages = "1"
	}

	return ResourceExhibitions{Data: resources, Links: links, Meta: Meta{TotalPages: totalPages}}
}


// http http://localhost:8080/api/exhibitions
func GetExhibitions(c *gin.Context) {
	var exhibitions []Exhibition
	page := c.DefaultQuery("page", "1")

	query := PaginateQuery("SELECT * FROM exhibiciones ORDER BY id", page)
	_, err := dbmap.Select(&exhibitions, query)
	checkErr(err)

	if err == nil {
		content := newResourceExhibitions(exhibitions, "exhibitions", page, "80")
		c.JSON(200, content)
	} else {
		jsonErr := &Error{"not_found", 404, "Not Found Error", "No exhibition found."}
		c.JSON(404, jsonErr)
	}
}

// http http://localhost:8080/api/exhibitions/1
func GetExhibition(c *gin.Context) {
	id := c.Params.ByName("e_id")
	var exhibition Exhibition
	err := dbmap.SelectOne(&exhibition, "SELECT * FROM exhibiciones WHERE id=? LIMIT 1", id)
	checkErr(err)

	if err == nil {
		content := newResourceExhibition(exhibition)
		c.JSON(200, content)
	} else {
		jsonErr := &Error{"not_found", 404, "Not Found Error", "Your query didn't match any exhibition."}
		c.JSON(404, jsonErr)
	}
}


api.go
package main

import (
	"os"
	"fmt"
	"log"
	//"net/http"
	//"path"
	"strconv"
 	"database/sql"
	"github.com/coopernurse/gorp"
	_ "github.com/mattn/go-sqlite3"
	"github.com/aviddiviner/gin-limit"
	"github.com/gin-gonic/gin"
)

var (
	//tmplDirectory string = "./tmpl"
	dbFilename    string = "./db/electronicArtArgentina.sqlite3"
	tableName     string = "exhibiciones"
	httpPort      string = "8080"
	apiHostname   string = "http://api.arte-electronico.cyberpunk.com.ar"
	apiPath       string = ""
	perPage       int = 20
	firstPage     string = "1"
	lastPage      string = "80" // (1593 / perPage) // TODO FIX ME

)

var dbmap = initDb()

type (
	PaginationLinks struct {
		First      string `json:"first"`
		Last       string `json:"last"`
		Prev       string `json:"prev"`
		Next       string `json:"next"`
	}

	Meta struct {
		TotalPages string `json:"total_pages"`
	}

	Error struct {
		Id     string `json:"id"`
		Status int `json:"status"`
		Title  string `json:"title"`
		Detail string `json:"detail"`
	}

)

func paginationLink(endpoint string, pg string, which string) string {
	page, err := strconv.ParseInt(pg, 10, 32)
	checkErr(err)
	first, ferr := strconv.ParseInt(firstPage, 10, 32)
	checkErr(ferr)
	last, lerr := strconv.ParseInt(lastPage, 10, 32)
	checkErr(lerr)


	switch which {
	case "prev":
		if page > first {
			page = page - 1
		}
	case "next":
		if page <= last {
			page = page + 1
		}
	}


	link := fmt.Sprintf("%s/%s?page=%d&perPage=%d", apiPath, endpoint, page, perPage)
	if endpoint == "search" && which == "last" { // TODO ugly
		link = ""
	}

	return link
}

func pageOffset(pg string) int {
	page, err := strconv.Atoi(pg)
	if err != nil {
		page = 1
	}

	return perPage * (page - 1)
}

func groupByQuery(query, groupBy string) string {
	field := ""
	order_by := "id"
	switch groupBy {
	case "event":
		field = "nombre_exhibicion"
	case "location":
		field = "nombre_locacion"
		order_by = "nombre_locacion"
	case "year":
		field = "anio"
		order_by = "anio"
	case "technique":
		field = "tecnica"
		order_by = "tecnica"
	}

	if field != "" {
		query = fmt.Sprintf(query, field, field)
		return fmt.Sprintf("%s GROUP BY %s ORDER BY %s", query, field, order_by)
	} else {
		return query
	}
}

func PaginateQuery(query, pg string) string {
	return fmt.Sprintf("%s LIMIT %d OFFSET %d", query, perPage, pageOffset(pg))
}

func init() {
	p := os.Getenv("HTTP_PORT")
	if p != "" {
		httpPort = p
	}

	d := os.Getenv("DB_FILENAME")
	if d != "" {
		dbFilename = d
	}

	// t:= os.Getenv("TMPL_DIR")
	// if t != "" {
	// 	tmplDirectory = t
	// }
}

func main() {
	router := gin.Default()
	router.Use(ApiCors())
	router.Use(limit.MaxAllowed(5))

	// API v1
	v1 := router.Group("v1")
	{
		v1.GET("/", ListEndpoints)  // list all exhibitions
		v1.GET("/exhibitions", GetExhibitions)  // list all exhibitions
		v1.GET("/exhibitions/:e_id", GetExhibition)  // show an exhibition
		v1.GET("/search", SearchExhibitions)  // search exhibitions

		v1.OPTIONS("/", EndpointsOptions) // http options
		v1.OPTIONS("/exhibitions", EndpointsOptions) // http options
		v1.OPTIONS("/exhibitions/:e_id", EndpointsOptions) // http options
		v1.OPTIONS("/search", EndpointsOptions) // http options
	}

	// API v2
	v2 := router.Group("v2")
	{
		v2.GET("/", ListEndpoints)  // list all exhibitions
		v2.GET("/exhibitions", GetExhibitions)  // list all exhibitions
		v2.GET("/exhibitions/:e_id", GetExhibition)  // show an exhibition
		v2.GET("/search", SearchExhibitions)  // search exhibitions
		v2.GET("/locations", GetLocations)  // list all exhibitions
		v2.GET("/events", GetEvents)  // list all exhibitions
		v2.GET("/techniques", GetTechniques)  // list all exhibitions

		v2.OPTIONS("/", EndpointsOptions) // http options
		v2.OPTIONS("/exhibitions", EndpointsOptions) // http options
		v2.OPTIONS("/exhibitions/:e_id", EndpointsOptions) // http options
		v2.OPTIONS("/search", EndpointsOptions) // http options
		v2.OPTIONS("/locations", EndpointsOptions) // http options
		v2.OPTIONS("/events", EndpointsOptions) // http options
		v2.OPTIONS("/techniques", EndpointsOptions) // http options
	}

	// API -- TODO improve this
	current := router.Group("/")
	{
		current.GET("/", ListEndpoints)  // list all exhibitions
		current.GET("/exhibitions", GetExhibitions)  // list all exhibitions
		current.GET("/exhibitions/:e_id", GetExhibition)  // show an exhibition
		current.GET("/search", SearchExhibitions)  // search exhibitions
		current.GET("/locations", GetLocations)  // list all exhibitions
		current.GET("/events", GetEvents)  // list all exhibitions
		current.GET("/techniques", GetTechniques)  // list all exhibitions

		current.OPTIONS("/", EndpointsOptions) // http options
		current.OPTIONS("/exhibitions", EndpointsOptions) // http options
		current.OPTIONS("/exhibitions/:e_id", EndpointsOptions) // http options
		current.OPTIONS("/search", EndpointsOptions) // http options
		current.OPTIONS("/locations", EndpointsOptions) // http options
		current.OPTIONS("/events", EndpointsOptions) // http options
		current.OPTIONS("/techniques", EndpointsOptions) // http options
	}

	// Home
	/*
	router.LoadHTMLGlob(path.Join(tmplDirectory, "*"))
	router.GET("/", func(c *gin.Context) {
		c.HTML(200, "index.tmpl", gin.H{
			"title": "JSON API for Dataset Arte Electrónico Argentino",
			"version": "v2",
		})
	})
	*/
	// router.GET("/", func(c *gin.Context) {})

	router.Run(fmt.Sprintf(":%s", httpPort))
}

func checkErr(err error) {
	if err != nil {
		log.Println(err)
	}
}

func initDb() *gorp.DbMap {
	db, err := sql.Open("sqlite3", dbFilename)
	checkErr(err)
	dbmap := &gorp.DbMap{Db: db, Dialect: gorp.SqliteDialect{}}
	dbmap.AddTableWithName(Exhibition{}, tableName).SetKeys(true, "Id")
	err = dbmap.CreateTablesIfNotExists()
	checkErr(err)
	return dbmap
}

func EndpointsOptions(c *gin.Context) {
	c.Writer.Header().Set("Access-Control-Allow-Methods", "OPTIONS, GET")
	c.Writer.Header().Set("Access-Control-Allow-Headers", "Content-Type")
	c.Next()
}

func ApiCors() gin.HandlerFunc {
	return func(c *gin.Context) {
		c.Writer.Header().Add("Access-Control-Allow-Origin", "*")
		c.Next()
	}
}

func ListEndpoints(c *gin.Context) {
	api_url := fmt.Sprintf("%s%s/", apiHostname, apiPath)
	c.JSON(200, gin.H{
		"techniques_url": fmt.Sprintf("%s%s{?page}", api_url, "techniques"),
		"events_url": fmt.Sprintf("%s%s{?page}", api_url, "events"),
		"locations_url": fmt.Sprintf("%s%s{?page}", api_url, "locations"),
		"exhibitions_url": fmt.Sprintf("%s%s{/exhibition_id}{?page}", api_url, "exhibitions"),
		"search_url": fmt.Sprintf("%s%s?q={query}{&year,when,technique,curator,artist,work,page}", api_url, "search"),
	})
}
